### PR TITLE
Support tso microservice auto service discovery in the pd client.

### PR DESCRIPTION
**Pull request by binshi-bing**
### What problem does this PR solve?

Issue Number: Ref #5836 

### What is changed and how does it work?

```commit-message
Support tso microservice auto service discovery in the pd client.
```

### Check List

Tests

- Unit test
- Manual test (add detailed scripts or steps below)
 ~/code/pingcap/my-pd/bin   tso-client-svc-discovery  ./pd-tso-bench -v -m -duration 10s -pd "127.0.0.1:2379"                                                                                                                                                                                       ✔  5354  16:29:52

Start benchmark #0, duration: 10s
Create 1 client(s) for benchmark
[2023/02/28 16:29:56.663 -08:00] [INFO] [client.go:306] ["[pd(tso)] create tso client with endpoints"] [pd(api)-address="[127.0.0.1:2379]"]
[2023/02/28 16:29:56.665 -08:00] [INFO] [base_client.go:477] ["[pd] switch leader"] [new-leader=http://127.0.0.1:2379] [old-leader=]
[2023/02/28 16:29:56.665 -08:00] [INFO] [base_client.go:153] ["[pd] init cluster id"] [cluster-id=7203525516143219003]
[2023/02/28 16:29:56.666 -08:00] [INFO] [tso_client.go:353] ["[pd(tso)] switch primary"] [new-primary=http://127.0.0.1:3380] [old-primary=]
[2023/02/28 16:29:56.666 -08:00] [INFO] [tso_request_dispatcher.go:287] ["[pd/tso] tso dispatcher created"] [dc-location=global]

count: 498504, max: 4.2795ms, min: 0.3284ms, avg: 2.0016ms
<1ms: 684, >1ms: 271848, >2ms: 225972, >5ms: 0, >10ms: 0, >30ms: 0, >50ms: 0, >100ms: 0, >200ms: 0, >400ms: 0, >800ms: 0, >1s: 0
count: 494935, max: 3.9645ms, min: 0.3201ms, avg: 2.0182ms
<1ms: 562, >1ms: 257241, >2ms: 237132, >5ms: 0, >10ms: 0, >30ms: 0, >50ms: 0, >100ms: 0, >200ms: 0, >400ms: 0, >800ms: 0, >1s: 0

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/tso_client.go [hunk 1]
```diff
@@ -89,15 +93,26 @@ func (c *client) GetLocalTSWithinKeyspaceAsync(ctx context.Context, dcLocation s
 	return req
 }
 
-var _ BaseClient = (*tsoBaseClient)(nil)
+const (
+	// tsoKeyspaceGroupPrimaryElectionPrefix defines the key prefix for keyspace group primary election.
+	// The entire key is in the format of "/pd/<cluster-id>/microservice/tso/keyspace-group-XXXXX/primary" in which
+	// XXXXX is 5 digits integer with leading zeros. For now we use 0 as the default cluster id.
+	tsoKeyspaceGroupPrimaryElectionPrefix = "/pd/0/microservice/tso/keyspace-group-"
```
Comment (rleungx) [inline client/tso_client.go:100]: I noticed that there is a comment left previously. See https://github.com/tikv/pd/pull/6008#discussion_r1118504891

File: client/tso_client.go [hunk 2]
```diff
@@ -296,6 +350,28 @@ func (c *tsoBaseClient) switchPrimary(addrs []string) error {
...
+	} else if resp.Count > 1 {
+		return errs.ErrClientGetMultiResponse.FastGenByArgs(resp.Kvs)
+	}
+
+	value := resp.Kvs[0].Value
+	member := &pdpb.Member{}
+	if err := proto.Unmarshal(value, member); err != nil {
+		return errs.ErrClientProtoUnmarshal.Wrap(err).GenWithStackByCause()
+	}
+	return c.switchPrimary(addrsToUrls([]string{member.Name}))
```
Comment (rleungx) [inline client/tso_client.go:376]: member.Name?

Comment (binshi-bing) [reply client/tso_client.go:376]: That's a hack previously by reusing pdpb.Member. My next pr will introduce a new struct as we discussed before.

File: client/tso_client.go (hunk 1 above)
Comment (binshi-bing) [reply client/tso_client.go:100]: Yes, we can certainly move "keyspace-group-" from the path to make it more compact. I'll make the change.

File: client/client.go [hunk 3]
```diff
@@ -290,21 +291,30 @@ func NewClientWithContext(ctx context.Context, svrAddrs []string, security Secur
...
 		return nil, err
 	}
 	return c, nil
 }
 
 // NewTSOClientWithContext creates a TSO client with context.
-func NewTSOClientWithContext(ctx context.Context, svrAddrs []string, security SecurityOption, opts ...ClientOption) (Client, error) {
-	log.Info("[pd(tso)] create tso client with endpoints", zap.Strings("tso-address", svrAddrs))
+// TODO:
+// relace NewClientWithContext with this API after we enable auto service discovery.
```
[the remark this answers is not in the document]
Comment (binshi-bing) [reply client/client.go:303]: changed locally.

**Comment by rleungx** [comment]
We need some basic tests to validate that it works, both the unit test and manual test are good.

**Review by rleungx — APPROVED** [review]

File: client/client.go [hunk 4]
```diff
@@ -290,21 +291,30 @@ func NewClientWithContext(ctx context.Context, svrAddrs []string, security Secur
 	c, clientCtx, clientCancel, tlsCfg := createClient(ctx, &security)
 	c.tsoStreamBuilderFactory = &pdTSOStreamBuilderFactory{}
 	c.bc = newPDBaseClient(clientCtx, clientCancel, &c.wg, addrsToUrls(svrAddrs), tlsCfg, c.option)
+	c.tsobc = c.bc
```
Comment (lhy1024) [inline client/client.go:294]: the same two clients?

**Review by lhy1024 — APPROVED** [review]
mostly LGTM

File: client/client.go (hunk 4 above)
Comment (binshi-bing) [reply client/client.go:294]: tsobc is an interface of tso baseclient which can be either tso in pd client or tso microservice client. For the former, tso in pd client is the pd client itself, that's why c.tsobc = c.bc. Actually, I planned one more refactor to make it much clearer -- Client contains pdServiceDiscovery by renaming pdBaseClient after we move all the tso related logic out of pdBaseClient. c.tsobc will be backed by either pdServiceDiscovery or tsoServiceDiscovery. After that, we won't see 'c.tsobc = c.bc' such thing which looks weird at the first sight.

**Pull request merged**